### PR TITLE
Check that linked issues are open

### LINKED ISSUE #2733: Make linter for checking that linked issues are open
### What should be done?

We have a linter that checks comments like `// TODO https://github.com/FerretDB/FerretDB/issues/<number>` in our Go code: https://github.com/FerretDB/FerretDB/tree/main/tools/checkcomments

It should also check if those issues are still open. Doing that is a bit tricky: the tool runs once for each Go package (not project or Go module) and generates so many requests (many of which are for the same issues) that we easily go over GitHub's rate limit. We should keep cache check results in a temporary file and coordinate multiple invocations using [`lockedfile`](https://github.com/rogpeppe/go-internal/tree/master/lockedfile).

Additionally (and in a separate PR), we should check those issue links in string constants like in
```go
skip: "https://github.com/FerretDB/FerretDB/issues/1522"
```

### Where?

https://github.com/FerretDB/FerretDB/tree/main/tools/checkcomments

### Definition of Done

* Linter does not check the same issue twice for one `task lint` invocation.
* String constants in Go code are checked.

### PR DESCRIPTION
**Pull request by KrishnaSindhur**
# Description

Closes #2733.

## Readiness checklist

- [ ] I added/updated unit tests (and they pass).
- [ ] I added/updated integration/compatibility tests (and they pass).
- [ ] I added/updated comments and checked rendering.
- [ ] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [ ] I ensured that PR title is good enough for the changelog.
- [ ] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [ ] I marked all done items in this checklist.

### CODE REVIEW

**Review by AlekSi — COMMENTED** [review]
Could you please add tests to testdata.go with an open and closed issue? If implemented correctly, they will show that this code doesn't detect closed issues correctly

**Comment by KrishnaSindhur** [comment]
@AlekSi I realised now there in no clear way to find issue open or closed by issue url like we used in our test, is there any better way?

**Comment by AlekSi** [comment]
I think we could use the GitHub API via https://github.com/google/go-github. We have some examples there: https://github.com/FerretDB/github-actions/blob/main/conform-pr/main.go

**Comment by AlekSi** [comment]
@KrishnaSindhur, any news? Anything we could help you with?

**Comment by KrishnaSindhur** [comment]
@AlekSi this PR is like draft just want to know how to test locally? github token is required to test it locally and some repo information please check!!

File: tools/checkcomments/checkcomments.go [hunk 1]
```diff
@@ -59,3 +66,67 @@ func run(pass *analysis.Pass) (any, error) {
...
+// isIssueopen check the issue open or closed
+func isIssueOpen(todoText string) bool {
+	issueURL := getURL(todoText)
+	if issueURL == "" {
+		return false
+	}
+
+	token := os.Getenv("GITHUB_TOKEN")
+	if token == "" {
+		log.Fatalf("%s is not set.", token)
```
Comment (AlekSi) [inline tools/checkcomments/checkcomments.go:79]: It is not actually required, let's remove it

**Review by AlekSi — COMMENTED** [review]
Tests fail because they are getting URLs like `https://api.github.com/repos/owner/https://github.com/FerretDB/FerretDB/issues/2733`. They don't look valid :)

**Comment by KrishnaSindhur** [comment]
what we should do if we have below URL? should we ignore or catch it? 
`https://api.github.com/repos/owner/https://github.com/FerretDB/FerretDB/issues/2733`

**Comment by AlekSi** [comment]
We should use construct valid URLs to check issues. For example, the given API URL for https://github.com/FerretDB/FerretDB/issues/2733 is https://api.github.com/repos/FerretDB/FerretDB/issues/2733

**Review by AlekSi — COMMENTED** [review]
Please run `task init fmt` and commit changes.

Please also add test comments to `checkcomments/testdata/testdata.go`.

**Review by AlekSi — COMMENTED** [review]
We should add those `// want` comments. They are actually tested by checkcomments_test.go. Try to set `// TODO` comment with a closed issue and `// want` with some invalid error message, run tests, and see them fail.

File: tools/checkcomments/testdata/testdata.go [hunk 2]
```diff
@@ -16,13 +16,21 @@
...
 }
 
-func testIncorrect() {
-	// TODO: https://github.com/FerretDB/FerretDB/issues/2733 // want "invalid TODO comment"
+func testIncorrectFormat() {
+	// TODO: https://github.com/FerretDB/FerretDB/issues/3413 // want "invalid TODO: incorrect format"
+}
+
+func testIncorrectClosed() {
+	// TODO https://github.com/FerretDB/FerretDB/issues/1 // want "invalid TODO: linked issue is closed"
```
Comment (AlekSi) [inline tools/checkcomments/testdata/testdata.go:31]: I updated tests to make error messages better. Please `git pull` and make them pass.

Comment (KrishnaSindhur) [reply tools/checkcomments/testdata/testdata.go:31]: @AlekSi how can i get the github token creds to run?

Comment (AlekSi) [reply tools/checkcomments/testdata/testdata.go:31]: They are not required. I can verify it yourself by running `curl https://api.github.com/repos/FerretDB/FerretDB/issues/2733` in the terminal without any credentials

**Review by AlekSi — COMMENTED** [review]
https://github.com/FerretDB/FerretDB/pull/3277#discussion_r1345202350

File: tools/checkcomments/testdata/testdata.go (hunk 2 above)
Comment (KrishnaSindhur) [reply tools/checkcomments/testdata/testdata.go:31]: updated to pass test this should work now!! as per test message `//want `

File: tools/checkcomments/checkcomments.go [hunk 3]
```diff
@@ -59,3 +70,71 @@ func run(pass *analysis.Pass) (any, error) {
...
+	token := ""
+	ctx := context.Background()
+	httpClient := oauth2.NewClient(ctx, oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	))
+
+	client := github.NewClient(httpClient)
+
+	owner := "owner"
+	repo := "repos"
```
Comment (AlekSi) [inline tools/checkcomments/checkcomments.go:90]:
```suggestion
	owner := "FerretDB"
	repo := "FerretDB"
```

`owner` and `repo` in the documentation are placeholders that should be replaced by repository owner (user or organization name) and repository name.

File: tools/checkcomments/checkcomments.go [hunk 4]
```diff
@@ -59,3 +70,71 @@ func run(pass *analysis.Pass) (any, error) {
...
+
+// extracting url from TODO comment if present.
+func getURL(todoText string) string {
+	arrText := strings.Split(todoText, " ")
+	for _, text := range arrText {
+		if strings.Contains(text, "https://") {
+			parts := strings.Split(text, "/")
+			if len(parts) >= 5 {
+				issueNumber := parts[6]
+				apiURL := fmt.Sprintf("https://api.github.com/repos/FerretDB/FerretDB/issues/%s", issueNumber)
```
Comment (AlekSi) [inline tools/checkcomments/checkcomments.go:113]: What are we doing there? We create a URL to check issue status that we don't actually use. Why?

Comment (KrishnaSindhur) [reply tools/checkcomments/checkcomments.go:113]: Yes correct we don't want that as creating url as we are using  github client to get the state value we can remove it!!

File: tools/checkcomments/checkcomments.go [hunk 5]
```diff
@@ -50,7 +57,13 @@ func run(pass *analysis.Pass) (any, error) {
 					}
 
 					if !todoRE.MatchString(c.Text) {
```
Comment (AlekSi) [inline tools/checkcomments/checkcomments.go:59]: `todoRE` should also be used to extract an issue number

File: tools/checkcomments/checkcomments.go [hunk 6]
```diff
@@ -59,3 +72,51 @@ func run(pass *analysis.Pass) (any, error) {
...
+		return false
+	}
+
+	return issue.GetState() == "open"
+}
+
+// get the issue number from  issue url.
+func getIssueNumber(todoText string) (int, error) {
+	pattern := `\/issues\/(\d+)`
+	re := regexp.MustCompile(pattern)
```
Comment (AlekSi) [inline tools/checkcomments/checkcomments.go:108]: `todoRE` already correctly extracts the issue number only for FerretDB/FerretDB repo. That code would accept issues from any repos. That's not what we want.

**Review by AlekSi — COMMENTED** [review]
I added code that uses `GITHUB_TOKEN` correctly. The linter now works with a token (that's primarily needed for CI) and without it.

And CI now correctly shows that error messages are not what is expected: https://github.com/FerretDB/FerretDB/actions/runs/6427704926/job/17453776093?pr=3277

**Comment by KrishnaSindhur** [comment]
@AlekSi below test fails because after TODO we have comment so it gives `invalid TODO: incorrect format` not the `invalid TODO: linked issue is closed` to get `invalid TODO: linked issue is closed` we have to remove `// want`
```
func testIncorrectClosed() {
	// TODO https://github.com/FerretDB/FerretDB/issues/1 // want "invalid TODO: linked issue is closed"
}
```

**Comment by AlekSi** [comment]
> we have to remove // want

That would be the equivalent of removing a test, and we want to keep that test. Let's find some other solution. For example, we could remove `// want …` part if the file name is `testdata.go`.

Let's also gather all URLs first, remove duplicates, and then check them. Right now we check the same issue multiple times, quickly running over a rate limit for unauthenticated requests.

**Review by AlekSi — COMMENTED** [review]
https://github.com/FerretDB/FerretDB/pull/3277#issuecomment-1752817130

**Comment by KrishnaSindhur** [comment]
@AlekSi should i make change to remove? this should be check in todo statement to check file name something and then avoid  // want ..
```
// want … part if the file name is testdata.go.
```
so then how do we check for duplicate issue without a api hit?

**Comment by AlekSi** [comment]
> so then how do we check for duplicate issue without a api hit?

By collecting already checked URLs and not checking the same URL twice

**Comment by KrishnaSindhur** [comment]
hey sorry i was occupied with some other things could not spend time last 2 weeks 
@AlekSi

**Comment by AlekSi** [comment]
No worries! :)

**Pull request merged**